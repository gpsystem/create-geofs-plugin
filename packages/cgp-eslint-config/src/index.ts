import merge from "lodash.merge";
import {
  eslintConfigBaseDependencies,
  eslintConfigBases,
  EslintConfigNames,
  UnknownEslintConfigBase,
} from "./configurations";

export type MoreThanOneArray<T> = [T, ...T[]];

// TODO: is there a type for a eslint configuration file?
/**
 * Merges eslint configuration templates into one configuration to output to a .eslintrc file.
 *
 * @param configNames All the configuration templates.
 * The templates will be merged (deeply) with further right configuration templates overwriting further left templates.
 *
 * @returns A tuple. First is the configuration, ready to be stringified and outputted to a .eslintrc file.
 * Second is the names of the dependencies to install to make sure the .eslintrc configuration works properly.
 */
export function getEslintConfig(
  ...configNames: MoreThanOneArray<EslintConfigNames>
): [config: Record<string, unknown>, dependencies: string[]] {
  if (!configNames.length)
    throw new Error("must provide at least one config name");
  const configsToMerge: UnknownEslintConfigBase[] = configNames.map(
    (val) => eslintConfigBases[val]
  );
  const config: Record<string, unknown> = merge(
    ...(configsToMerge as MoreThanOneArray<UnknownEslintConfigBase>)
  );

  // if the typescript template is mentioned, remove globals from the config
  if (configNames.includes("tsBase") && config.globals !== undefined)
    delete config.globals;

  return [
    config,
    configNames.reduce<string[]>(
      (acc, val) => [...acc, ...eslintConfigBaseDependencies[val]],
      []
    ),
  ];
}

export { EslintConfigNames };
