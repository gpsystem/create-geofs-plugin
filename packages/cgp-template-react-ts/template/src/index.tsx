/// <reference types="jquery" />
/// <reference types="@geps/geofs-types" />

import React from "react";
import { createRoot } from "react-dom/client";
import { App } from "./App";

// Replace the selector with a link to your UI host!
// Example: #geofs-ui-right
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const root = createRoot(document.querySelector("REPLACE ME")!);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
