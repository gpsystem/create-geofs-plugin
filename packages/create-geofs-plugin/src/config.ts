import { resolve } from "node:path";
import { cwd } from "node:process";
import { Command } from "commander";
import { sync as packageDirectorySync } from "pkg-dir";
import { version } from "../package.json";

export interface Configuration {
  targetDir: string;
  templateName: "default" | "typescript" | "react" | "react-ts";
  templateDir: string;
  overwrite: boolean;
}

export function parseConfig(argv: string[]): Configuration {
  let destination = "";

  // eslint-disable-next-line prefer-const
  let { overwrite, template } = new Command()
    .name("create-geofs-plugin")
    .version(`Create GeoFS Plugin v${version}`)
    .argument("<destination>")
    .action((dest: string) => {
      if (!dest.includes("/"))
        throw new Error("argument <destination> must be a relative path");
      else destination = resolve(cwd(), dest);
    })
    .option("--overwrite", "Overwrite existing files", false)
    .option("-t, --template <template>", "Name of the template to use")
    .parse(argv)
    .opts<{ overwrite?: boolean; template?: string }>();

  if (!["default", "typescript", "react", "react-ts"].includes(template ?? ""))
    template = "default";

  // TODO: create templates
  const templateDir: string | undefined = require.resolve(
    // `@geps/cgp-template-${template}`
    `@geps/cgp-eslint-config`
  );
  if (!templateDir) throw new Error("template not found");

  return {
    targetDir: destination,
    overwrite: overwrite ?? false,
    templateName: template as "default" | "typescript" | "react" | "react-ts",
    templateDir,
  };
}
