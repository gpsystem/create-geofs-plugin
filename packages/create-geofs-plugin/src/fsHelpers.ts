import {
  Dirent,
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";

function pathIsDirectory(path: string): boolean {
  return statSync(path).isDirectory();
}

function internalWriteFileSync(targetFile: string, data: string): void {
  writeFileSync(targetFile, data, "utf-8");
}

export function outputFile(
  targetFile: string,
  data: string,
  force = false
): void {
  if (!force && existsSync(targetFile)) throw new Error("file already exists");

  const targetFileDirectory: string = dirname(targetFile);

  if (!existsSync(targetFileDirectory))
    mkdirSync(targetFileDirectory, { recursive: true });
  internalWriteFileSync(targetFile, data);
}

// TODO: should we make a wrapper for this that throws on initialization and returns the full string[]?
export function* getAllFilesInDir(
  dirName: string
): Generator<string, void, void> {
  if (!pathIsDirectory(dirName))
    throw new Error(`passed path is not a directory`);

  const directoryDetails: Dirent[] = readdirSync(dirName, {
    withFileTypes: true,
  });

  for (const directoryChild of directoryDetails) {
    const childPath: string = resolve(dirName, directoryChild.name);

    if (directoryChild.isDirectory()) yield* getAllFilesInDir(childPath);
    else if (directoryChild.isFile()) yield childPath;
  }
}

export interface CopyDirOptions {
  overwrite?: boolean;
}

export function copyDir(
  srcDir: string,
  targetDir: string,
  { overwrite = false }: CopyDirOptions = {}
): void {
  if (srcDir === targetDir)
    throw new Error("source and target directories can't be the same");
  if (!pathIsDirectory(srcDir))
    throw new Error("source path isn't a directory");

  for (const srcFilePath of getAllFilesInDir(srcDir)) {
    const newFilePath: string = join(targetDir, relative(srcDir, srcFilePath));

    if (existsSync(newFilePath) && overwrite === false) continue;
    else outputFile(newFilePath, readFileSync(srcFilePath, "utf-8"), overwrite);
  }
}
