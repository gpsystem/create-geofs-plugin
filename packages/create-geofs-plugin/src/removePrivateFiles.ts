export default function removePrivateFiles(directoryPath: string) {
  // all private files start with __
}
