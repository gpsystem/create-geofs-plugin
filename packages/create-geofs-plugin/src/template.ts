import { readFileSync } from "node:fs";
import { EslintConfigNames, getEslintConfig } from "@geps/cgp-eslint-config";

export interface TemplateJson {
  /** The dependencies of the template. Does not include the dependencies returned from cgp-eslint-config. */
  dependencies: string[];
  eslintConfigTemplates: EslintConfigNames[];
}

function isObject(obj: unknown): obj is Record<string, unknown> {
  // intentionally allowing a comparison of both undefined and null
  // Object.getPrototypeOf will throw for both undefined and null
  // eslint-disable-next-line eqeqeq
  return obj != null && Object.getPrototypeOf(obj) === Object.prototype;
}

export function assertTemplateJsonFormat(
  templateJson: unknown
): templateJson is TemplateJson {
  // thanks to Joshua Chen on the TS discord for this assertion code!
  type AssertTrue<T extends true> = T;
  // If typescript throws type 'false' does not satisfy the constraint 'true' here, then
  // allPossibleConfigNames is malformed
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  type A = AssertTrue<
    [EslintConfigNames] extends [typeof allPossibleConfigNames[number]]
      ? true
      : false
  >;

  const allPossibleConfigNames = ["baseConfig", "tsBase", "reactBase"] as const;
  const { isArray } = Array;

  return (
    isObject(templateJson) &&
    isArray(templateJson.dependencies) &&
    templateJson.dependencies.every<string>((val): val is string => {
      // TODO: ensure val is a npm dep name followed by a major version number (after @ symbol)
      return typeof val === "string";
    }) &&
    isArray(templateJson.eslintConfigTemplates) &&
    templateJson.eslintConfigTemplates.every<EslintConfigNames>(
      (val): val is EslintConfigNames => allPossibleConfigNames.includes(val)
    )
  );
}

export interface ExpandedTemplateJson {
  /** All dependencies of the template, including the eslint dependencies. */
  dependencies: string[];
  /** The configuration to save to .eslintrc.yml. */
  eslintConfig: Record<string, unknown>;
}

export default function expandTemplateJson(
  templateJsonPath: string
): ExpandedTemplateJson {
  const templateJson: unknown = JSON.parse(
    readFileSync(templateJsonPath, "utf-8")
  );

  if (!assertTemplateJsonFormat(templateJson))
    throw new Error("template.json is malformed");

  const { eslintConfigTemplates } = templateJson;
  if (eslintConfigTemplates.length === 0)
    throw new Error("template must specify at least one config template");
  if (eslintConfigTemplates[0] !== "baseConfig")
    throw new Error(
      "template's eslint config template list must include the base configuration first"
    );

  const [eslintConfig, eslintDependencies] = getEslintConfig(
    ...(eslintConfigTemplates as [EslintConfigNames, ...EslintConfigNames[]])
  );

  const dependencies: string[] = [
    ...templateJson.dependencies,
    ...eslintDependencies,
  ];

  return {
    dependencies,
    // TODO
    eslintConfig: eslintConfig as Record<string, unknown>,
  };
}
