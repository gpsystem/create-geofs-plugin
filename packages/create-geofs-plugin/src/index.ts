import { join } from "node:path";
import { dir, DirectoryResult } from "tmp-promise";
import { Configuration, parseConfig } from "./config";
import { copyDir } from "./fsHelpers";
import expandTemplateJson, { ExpandedTemplateJson } from "./template";

export async function start(argv: string[]): Promise<void> {
  const config: Configuration = parseConfig(argv);
  const { path: tmpDirName, cleanup: cleanupTempDir }: DirectoryResult =
    await dir({
      unsafeCleanup: true,
    });

  copyDir(config.templateDir, tmpDirName, { overwrite: config.overwrite });
  const { dependencies: templateDeps, eslintConfig }: ExpandedTemplateJson =
    expandTemplateJson(join(tmpDirName, "template.json"));
  // remove everything that starts with __
  // copy everything to the target dir
  // install the dependencies
  await cleanupTempDir();
}
