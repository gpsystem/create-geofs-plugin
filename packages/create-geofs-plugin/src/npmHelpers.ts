import { execSync } from "node:child_process";
import { join } from "node:path";
import { outputFile } from "./fsHelpers";

export function initializeNpm(
  targetDir: string,
  templateDeps: string[],
  ci: boolean
): void {
  const firstPackageJson: Record<string, unknown> = {
    name: "geofs-plugin",
    version: "",
    private: true,
  };

  outputFile(
    join(targetDir, "package.json"),
    JSON.stringify(firstPackageJson, null, 2),
    {
      force: true,
    }
  );
  if (!ci) execSync(`npm i ${templateDeps.join(" ")} -D`, { cwd: targetDir });
}
