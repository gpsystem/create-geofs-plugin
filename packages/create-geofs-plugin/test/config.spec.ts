import { relative } from "node:path";
import { Configuration, parseConfig } from "../src/config";
import {
  allArgPermutations,
  createArgsForProgram,
  createTestDirectory__RENAME_LATER,
  normalizeToForwardSlash,
  TestDirectoryResults,
} from "./utils/index";

describe("parse config", () => {
  const { cleanup, directoryPath, getRelativePath }: TestDirectoryResults =
    createTestDirectory__RENAME_LATER();

  afterEach(() => cleanup());

  for (const [name, args] of allArgPermutations) {
    test(`ensure parsing integrity ${name}`, () => {
      const originalConfig: Configuration = parseConfig(
        createArgsForProgram(args, directoryPath)
      );
      const snapshotConfig: Configuration = {
        ...originalConfig,
        // should be an empty string
        targetDir: normalizeToForwardSlash(
          getRelativePath(originalConfig.targetDir)
        ),
        templateDir: normalizeToForwardSlash(
          relative(__dirname, originalConfig.templateDir)
        ),
      };

      expect(snapshotConfig).toMatchSnapshot();
    });
  }
});
