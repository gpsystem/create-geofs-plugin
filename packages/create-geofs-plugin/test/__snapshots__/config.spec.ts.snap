// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse config ensure parsing integrity overwrite:false 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-default",
  "templateName": "default",
}
`;

exports[`parse config ensure parsing integrity overwrite:false/template:react 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-react",
  "templateName": "react",
}
`;

exports[`parse config ensure parsing integrity overwrite:false/template:react-ts 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-react-ts",
  "templateName": "react-ts",
}
`;

exports[`parse config ensure parsing integrity overwrite:false/template:typescript 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-typescript",
  "templateName": "typescript",
}
`;

exports[`parse config ensure parsing integrity overwrite:true 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../cgp-template-default",
  "templateName": "default",
}
`;

exports[`parse config ensure parsing integrity overwrite:true/template:react 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../cgp-template-react",
  "templateName": "react",
}
`;

exports[`parse config ensure parsing integrity overwrite:true/template:react-ts 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../cgp-template-react-ts",
  "templateName": "react-ts",
}
`;

exports[`parse config ensure parsing integrity overwrite:true/template:typescript 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../cgp-template-typescript",
  "templateName": "typescript",
}
`;

exports[`parse config ensure parsing integrity template:react 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-react",
  "templateName": "react",
}
`;

exports[`parse config ensure parsing integrity template:react-ts 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-react-ts",
  "templateName": "react-ts",
}
`;

exports[`parse config ensure parsing integrity template:typescript 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-typescript",
  "templateName": "typescript",
}
`;

exports[`parse config ensure parsing integrity vanilla 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../cgp-template-default",
  "templateName": "default",
}
`;
